jax/interpreters/sharded_jit.py

```diff
@@ -13,262 +13,172 @@
-  subc._builder.set_sharding(_sharding_to_proto(partitions[1]))
-  out_tuple = subc.Tuple(*out_nodes)
-  subc._builder.clear_sharding()
+  if partitions is None:
+    return pxla.ShardingSpec(
+        # int(1) because pytype is confused by 1 (???)
+        shards_per_axis=(int(1),) * len(aval.shape),
+        is_axis_materialized=(True,) * len(aval.shape),
+        replication_factor=num_partitions)
+  else:
+    assert len(partitions) == len(aval.shape)
+    return pxla.ShardingSpec(
+        shards_per_axis=tuple(partitions),
+        is_axis_materialized=(True,) * len(aval.shape),
+        replication_factor=1)
 
-  subc = subc.Build(out_tuple)
-  return c.Call(subc, list(freevar_nodes) + list(in_nodes))
 
 def _execute_spatially_partitioned(compiled, in_handler, out_handler, *args):
   input_bufs = in_handler(args)
   out_bufs = compiled.execute_on_local_devices(list(input_bufs))
   return out_handler(out_bufs)
 
 
-def _xla_sharded_args(c, avals, partitions):
+def _xla_sharded_args(c, avals, in_parts):
   xla_args = []
-  for p, a in zip(partitions, avals):
-    c._builder.set_sharding(_sharding_to_proto(p))
-    # logging.error("===== aval shape: %s" % str(a.shape))
-    shape = xc.Shape.array_shape(a.dtype, (4,8))
-    xla_args.append(c.ParameterWithShape(xla.aval_to_xla_shape(a)))
-    c._builder.clear_sharding()
+  for i, (sharding, aval) in enumerate(safe_zip(in_parts, avals)):
+    param = xb.with_sharding(c, sharding, xb.parameter, c, i,
+                             xla.aval_to_xla_shape(aval))
+    xla_args.append(param)
   return xla_args
 
 
-def _sharding_to_proto(sharding):
-  proto = xc.OpSharding()
-  if isinstance(sharding, tuple):
-    if sharding[0] is None or isinstance(sharding[0], tuple):
-      sub_protos = [_sharding_to_proto(s) for s in sharding]
-      xc.type = xc.OpSharding.Type.TUPLE
-      xc.tuple_shardings = sub_protos
-      return proto
-
-  if sharding is None:
-    proto.type = xc.OpSharding.Type.REPLICATED
-  else:
-    proto.type = xc.OpSharding.Type.OTHER
-    proto.tile_assignment_dimensions = list(sharding)
-    proto.tile_assignment_devices = list(range(onp.product(sharding)))
-  return proto
-
-
-def _get_num_partitions(partitions):
-  num_partitions = onp.prod(onp.max(partitions, axis=0))
-  return num_partitions
-
-
-def get_num_partitions(partitions):
-  num_partitions_set = set(_get_num_partitions(parts) for parts in partitions)
+def get_num_partitions(*partitions):
+  partition_specs = tree_flatten(partitions)[0]
+  if len(partition_specs) == 0:
+    # Everything is specified as replicated (all Nones).
+    return 1
+  num_partitions_set = set(np.prod(spec) for spec in partition_specs)
   if len(num_partitions_set) > 1:
     raise ValueError(
-        "All partition specs must use the same number of total partitions, "
-        "got: %s" % partitions)
+        f"All partition specs must use the same number of total partitions, "
+        "got {partitions}, with distinct number of partitions "
```

Comment:
You have to use "f" before each string literal that contains interpolation in an implicit concatenation (I believe)

Reply:
Done.